src/generator/generator_visitor.ts

```diff
@@ -100,42 +103,86 @@ class GeneratorVisitor extends Visitor {
     protected visitBinaryExpression(node: BinaryExpression) {
         super.visitBinaryExpression(node);
 
-        const right = this.popExpression();
-        const left = this.popExpression();
+        let right = this.popExpression();
+        let left = this.popExpression();
+
+        const rightType = this.getExpressionType(node.right);
+        const leftType = this.getExpressionType(node.left);
+        const commonNumberType = getCommonNumberType(leftType, rightType);
+
+        if (leftType !== commonNumberType) {
+            left = this.module.f64.convert_s.i32(left);
+        }
+
+        if (rightType !== commonNumberType) {
+            right = this.module.f64.convert_s.i32(right);
+        }
+
+        const operationsInstance = this.getOperationsInstance(commonNumberType);
 
         switch (node.operator) {
             case '+':
-                this.expressions.push(this.module.i32.add(left, right));
+                this.expressions.push(operationsInstance.add(left, right));
                 break;
             case '-':
-                this.expressions.push(this.module.i32.sub(left, right));
+                this.expressions.push(operationsInstance.sub(left, right));
                 break;
             case '*':
-                this.expressions.push(this.module.i32.mul(left, right));
+                this.expressions.push(operationsInstance.mul(left, right));
                 break;
             case '/':
-                this.expressions.push(this.module.i32.div_s(left, right));
+                if (commonNumberType === WebAssemblyType.INT_32) {
+                    this.expressions.push(this.module.i32.div_s(left, right));
```

Comment:
Yes this is a different behavior than in JS. But I think that because of the reason that wasm is strongly typed, this behavior is and should be expected. It is, after all, a different language.

Good point, I don't think this is documented yet. I will add it when I'm describing the double type.